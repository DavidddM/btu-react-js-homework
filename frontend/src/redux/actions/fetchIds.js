import { startFetching, fetchIdsError, fetchIdsSuccess } from "../actions";
import settings from "../../settings";

function fetchIds(uid, jwt) {
    return (dispatch) => {
        dispatch(startFetching());
        fetch(`${settings.APP_BASE_URL}/users/${uid}`, {
            headers: {
                authorization: `Bearer ${jwt}`,
            },
        })
            .then((res) => res.json())
            .then((res) => {
                if (res.error) {
                    throw res.error;
                }
                dispatch(
                    fetchIdsSuccess({
                        ratedIds: res.links.map((l) => String(l.id)),
                    })
                );
            })
            .catch((error) => {
                dispatch(fetchIdsError({ fetchingErr: error }));
            });
    };
}

export default fetchIds;
