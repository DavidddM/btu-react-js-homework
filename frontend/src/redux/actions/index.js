import {
    INCREASE_RATING,
    DECREASE_RATING,
    SET_VALUE,
    SET_RATING_RIGHT,
    SET_IS_AUTH,
    SET_UID,
    SET_LOGIN_ERROR,
    ADD_RATED_ID,
    START_FETCHING_IDS,
    FETCHING_IDS_ERR,
    FETCHING_IDS_SUCCESS,
} from "../constants/action-types";

export function increaseRating() {
    return {
        type: INCREASE_RATING,
    };
}

export function decreaseRating() {
    return {
        type: DECREASE_RATING,
    };
}

export function setValue(payload) {
    return {
        type: SET_VALUE,
        payload: payload,
    };
}

export function setRatingRight(payload) {
    return {
        type: SET_RATING_RIGHT,
        payload: payload,
    };
}
export function setIsAuth(payload) {
    return {
        type: SET_IS_AUTH,
        payload: payload,
    };
}

export function setUid(payload) {
    return {
        type: SET_UID,
        payload: payload,
    };
}

export function setLoginError(payload) {
    return {
        type: SET_LOGIN_ERROR,
        payload: payload,
    };
}

export function addRatedId(payload) {
    return {
        type: ADD_RATED_ID,
        payload: payload,
    };
}

export function startFetching() {
    return {
        type: START_FETCHING_IDS,
    };
}

export function fetchIdsError(payload) {
    return {
        type: FETCHING_IDS_ERR,
        payload: payload,
    };
}

export function fetchIdsSuccess(payload) {
    return {
        type: FETCHING_IDS_SUCCESS,
        payload: payload,
    };
}
