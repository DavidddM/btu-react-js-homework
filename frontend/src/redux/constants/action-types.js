export const INCREASE_RATING = "INCREASE_RATING";
export const DECREASE_RATING = "DECREASE_RATING";
export const SET_VALUE = "SET_VALUE";
export const SET_RATING_RIGHT = "SET_RATING_RIGHT";
export const SET_IS_AUTH = "SET_IS_AUTH";
export const SET_LOGIN_ERROR = "SET_LOGIN_ERROR";
export const ADD_RATED_ID = "ADD_RATED_ID";
export const START_FETCHING_IDS = "START_FETCHING_IDS";
export const FETCHING_IDS_ERR = "FETCHING_IDS_ERR";
export const FETCHING_IDS_SUCCESS = "FETCHING_IDS_SUCCESS";
