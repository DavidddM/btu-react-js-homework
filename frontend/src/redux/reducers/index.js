import {
    INCREASE_RATING,
    DECREASE_RATING,
    SET_VALUE,
    SET_RATING_RIGHT,
    SET_IS_AUTH,
    SET_LOGIN_ERROR,
} from "../constants/action-types";

const initalState = {
    rating: 0,
    ratingRight: true,
    isAuth: false,
    loginError: false,
};

function rootReducer(state = initalState, action) {
    console.log(action);
    switch (action.type) {
        default:
            return state;
        case INCREASE_RATING:
            return {
                ...state,
                rating: ++state.rating,
            };

        case DECREASE_RATING:
            return {
                ...state,
                rating: --state.rating,
            };
        case SET_VALUE:
            return {
                ...state,
                rating: action.payload.rating,
            };
        case SET_RATING_RIGHT:
            return {
                ...state,
                ratingRight: action.payload.ratingRight,
            };
        case SET_IS_AUTH:
            return {
                ...state,
                isAuth: action.payload.isAuth,
            };
        case SET_LOGIN_ERROR:
            return {
                ...state,
                loginError: action.payload.loginError,
            };
    }
}

export default rootReducer;
