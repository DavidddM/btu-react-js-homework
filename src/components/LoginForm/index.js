import React, { useRef, useContext } from "react";
import { useForm } from "react-hook-form";

import { AuthContext } from "../../context/authContext";

function LoginForm() {
  const { login } = useContext(AuthContext);
  const { handleSubmit, register } = useForm();

  const onSubmit = (data) => {
    if (data.login === data.password && data.password === "admin") {
      login();
    }
  };

  const formRef = useRef(null);

  return (
    <div className="login-form">
      <form method="post" onSubmit={handleSubmit(onSubmit)} ref={formRef}>
        <h2 className="text-center">Log in</h2>
        <div className="form-group">
          <input
            type="text"
            className="form-control"
            placeholder="Username"
            required="required"
            name="login"
            id="login"
            ref={register}
          />
        </div>
        <div className="form-group">
          <input
            type="password"
            className="form-control"
            placeholder="Password"
            required="required"
            name="password"
            id="password"
            ref={register}
          />
        </div>
        <div className="form-group">
          <button type="submit" className="btn btn-primary btn-block">
            Log in
          </button>
        </div>
      </form>
    </div>
  );
}

export default LoginForm;
