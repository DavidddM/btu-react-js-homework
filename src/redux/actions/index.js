import {
    SET_IS_AUTH,
    SET_LOGIN_ERROR,
    SET_USERNAME,
} from "../constants/action_types";

export function setIsAuth(payload) {
    return {
        type: SET_IS_AUTH,
        payload: payload,
    };
}

export function setUsername(payload) {
    return {
        type: SET_USERNAME,
        payload: payload,
    };
}

export function setLoginError(payload) {
    return {
        type: SET_LOGIN_ERROR,
        payload: payload,
    };
}
