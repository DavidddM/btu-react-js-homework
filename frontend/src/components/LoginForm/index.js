import React, { useRef, useEffect } from "react";
import { useForm } from "react-hook-form";

import { connect } from "react-redux";
import { setIsAuth, setLoginError } from "../../redux/actions";
import axios from "axios";

import settings from "../../settings";

import { useHistory } from "react-router-dom";

function LoginForm(props) {
    const { handleSubmit, register } = useForm();
    const history = useHistory();

    useEffect(() => {
        props.setLoginError({ loginError: false });
    }, []);

    const onSubmit = (data) => {
        axios
            .post(`${settings.APP_BASE_URL}/auth/local`, {
                identifier: data.login,
                password: data.password,
            })
            .then((resp) => {
                localStorage.setItem("userName", resp.data.user.username);
                localStorage.setItem("token", resp.data.jwt);
                props.setIsAuth({ isAuth: true });
                history.push("/");
            })
            .catch((err) => {
                props.setLoginError({ loginError: true });
                formRef.current.elements.login.value = formRef.current.elements.password.value =
                    "";
            });
    };

    const formRef = useRef(null);

    return (
        <div
            className="login-form"
            style={{ width: "350px", margin: "0 auto" }}
        >
            <form method="post" onSubmit={handleSubmit(onSubmit)} ref={formRef}>
                <h2 className="text-center">Log in</h2>
                <div className="form-group">
                    <input
                        type="text"
                        className="form-control"
                        placeholder="Username"
                        required="required"
                        name="login"
                        id="login"
                        ref={register}
                    />
                </div>
                <div className="form-group">
                    <input
                        type="password"
                        className="form-control"
                        placeholder="Password"
                        required="required"
                        name="password"
                        id="password"
                        ref={register}
                    />
                </div>
                <div className="form-group">
                    <button type="submit" className="btn btn-primary btn-block">
                        Log in
                    </button>
                </div>
            </form>
            {props.loginError && (
                <div
                    style={{
                        margin: "0 auto",
                        textAlign: "center",
                    }}
                    className="alert alert-danger"
                    role="alert"
                >
                    Wrong login/password!
                </div>
            )}
        </div>
    );
}

const mapStateToProps = (state) => ({
    isAuth: state.isAuth,
    loginError: state.loginError,
    jwt: state.jwt,
});

const mapDispatchToProps = (dispatch) => {
    return {
        setIsAuth(payload) {
            dispatch(setIsAuth(payload));
        },
        setLoginError(payload) {
            dispatch(setLoginError(payload));
        },
        setJwt(payload) {
            dispatch(setJwt(payload));
        },
    };
};

const ConnectedLoginForm = connect(
    mapStateToProps,
    mapDispatchToProps
)(LoginForm);

export default ConnectedLoginForm;
