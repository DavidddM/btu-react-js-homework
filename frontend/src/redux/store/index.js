import { createStore, compose, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import rootReducer, { loginReducer, dataReducer } from "../reducers/index";

const combined = combineReducers({
    app: rootReducer,
    loginInfo: loginReducer,
    data: dataReducer,
});

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(combined, composeEnhancers(applyMiddleware(thunk)));

export default store;
