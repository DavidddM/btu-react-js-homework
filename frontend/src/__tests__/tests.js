import { describe, expect, it } from "@jest/globals";
import * as types from "../redux/constants/action-types";
import * as actions from "../redux/actions/";
import rootReducer, { loginReducer, dataReducer } from "../redux/reducers";
import React from "react";
import { configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import LoginForm from "../components/LoginForm";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

// მთლად ყველაფერს არ გავტესტავ, დიდ დროს წაიღებს და literally copy-paste იქნება

describe("root reducer", () => {
    it("should return the initial state", () => {
        expect(rootReducer(undefined, {})).toEqual({
            rating: 0,
            ratingRight: true,
        });
    });
    it("should return the rating: 15", () => {
        expect(
            rootReducer([], {
                type: types.SET_VALUE,
                payload: { rating: 15 },
            })
        ).toEqual({
            rating: 15,
        });
    });
});

describe("login reducer", () => {
    it("should return the initial state", () => {
        expect(loginReducer(undefined, {})).toEqual({
            uid: false,
            isAuth: false,
            loginError: false,
            jwt: false,
            userName: false,
        });
    });
    it("should return the isAuth: true", () => {
        expect(
            loginReducer([], {
                type: types.SET_IS_AUTH,
                payload: { isAuth: true },
            })
        ).toEqual({
            isAuth: true,
        });
    });
    it("should return the userName: David", () => {
        expect(
            loginReducer([], {
                type: types.SET_USERNAME,
                payload: { userName: "David" },
            })
        ).toEqual({
            userName: "David",
        });
    });
});

describe("data reducer", () => {
    it("should return the initial state", () => {
        expect(dataReducer(undefined, {})).toEqual({
            ratedIds: [],
            fetchPending: false,
            error: false,
        });
    });
    it("should return the fetchPending: true", () => {
        expect(
            dataReducer([], {
                type: types.START_FETCHING_IDS,
                payload: { fetchPending: true },
            })
        ).toEqual({
            fetchPending: true,
        });
    });
    it("should return the rateIds & fetchPending: false", () => {
        expect(
            dataReducer([], {
                type: types.FETCHING_IDS_SUCCESS,
                payload: { fetchPending: false, ratedIds: [1, 2, 3, 4] },
            })
        ).toEqual({
            fetchPending: false,
            ratedIds: [1, 2, 3, 4],
        });
    });
});

describe("login actions", () => {
    it("set isAuth to true", () => {
        const payload = {
            isAuth: true,
        };
        const expectedAction = {
            type: types.SET_IS_AUTH,
            payload,
        };
        expect(actions.setIsAuth(payload)).toEqual(expectedAction);
    });
    it("set random generated jwt", () => {
        const jwt = Math.floor(Math.random() * 1000000000000);
        const payload = {
            jwt: jwt,
        };
        const expectedAction = {
            type: types.SET_JWT,
            payload,
        };
        expect(actions.setJwt(payload)).toEqual(expectedAction);
    });
    it("set random username", () => {
        const payload = {
            userName: "david",
        };
        const expectedAction = {
            type: types.SET_USERNAME,
            payload,
        };
        expect(actions.setUsername(payload)).toEqual(expectedAction);
    });
});

const mockStore = configureStore([]);
configure({ adapter: new Adapter() });

describe("Login component", () => {
    let store;

    beforeEach(() => {
        store = mockStore({
            loginInfo: {
                uid: false,
                isAuth: false,
                loginError: false,
                jwt: false,
                userName: false,
            },
            app: {
                rating: 15,
                ratingRight: false,
            },
            data: {
                ratedIds: [1, 2, 3],
            },
        });
    });

    it("render LoginForm", () => {
        const component = mount(
            <Provider store={store}>
                <LoginForm />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });
});
