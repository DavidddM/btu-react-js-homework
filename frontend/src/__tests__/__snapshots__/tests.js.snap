// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component render LoginForm 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(LoginForm)>
    <LoginForm
      fetchIds={[Function]}
      loginError={false}
      setIsAuth={[Function]}
      setJwt={[Function]}
      setLoginError={[Function]}
      setUid={[Function]}
      setUsername={[Function]}
    >
      <div
        className="login-form"
        style={
          Object {
            "margin": "0 auto",
            "width": "350px",
          }
        }
      >
        <form
          method="post"
          onSubmit={[Function]}
        >
          <h2
            className="text-center"
          >
            Log in
          </h2>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              id="login"
              name="login"
              placeholder="Username"
              required="required"
              type="text"
            />
          </div>
          <div
            className="form-group"
          >
            <input
              className="form-control"
              id="password"
              name="password"
              placeholder="Password"
              required="required"
              type="password"
            />
          </div>
          <div
            className="form-group"
          >
            <button
              className="btn btn-primary btn-block"
              type="submit"
            >
              Log in
            </button>
          </div>
        </form>
      </div>
    </LoginForm>
  </Connect(LoginForm)>
</Provider>
`;
